### PR TITLE
Add facet types and spawn logic for sparsely-valued spawn heights and sea level heights.

### PR DESCRIPTION
**Pull request by PS-Soundwave**
Closes #3177

### Contains

Adds two new facet types, that allow spawn heights and sea level heights to not be defined for all XZ coordinates. If a spawn height is not defined for a coordinate, the player cannot spawn there. If a sea level height is not defined for a coordinate, the player can spawn there, regardless of height.

If there is no spawn height for which the player is above sea level, the player will be spawned without regard to sea level. It's still possible for there to not be a valid spawn location (e.g. no spawn heights specified), in which case this is considered a world generation error, and an exception is thrown.

### How to test

- Ensure that this PR remains backwards compatible with other world generators, which should remain unaffected in terms of spawn logic

- Use the new facets, and ensure that spawn locations are being chosen as expected

### CODE REVIEW

File: engine/src/main/java/org/terasology/world/generation/facets/base/StrictlySparseFieldFacet2D.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * limitations under the License.
+ */
+package org.terasology.world.generation.facets.base;
+
+import org.terasology.math.geom.BaseVector2i;
+import org.terasology.world.generation.WorldFacet2D;
+
+import java.util.Optional;
+
+public interface StrictlySparseFieldFacet2D extends WorldFacet2D {
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/world/generation/facets/base/StrictlySparseFieldFacet2D.java]: Again, either javadoc on the Base version of this class, or this interface. Whichever is consistent with the other world gen systems

File: engine/src/main/java/org/terasology/world/generation/facets/base/BaseStrictlySparseFieldFacet2D.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.Collections;
+import java.util.Optional;
+
+public class BaseStrictlySparseFieldFacet2D extends BaseFacet2D implements StrictlySparseFieldFacet2D {
+    private ArrayList<Optional<Float>> data;
+
+    public BaseStrictlySparseFieldFacet2D(Region3i targetRegion, Border3D border) {
+        super(targetRegion, border);
+        Vector2i size = getRelativeRegion().size();
+        this.data = new ArrayList<>(Collections.nCopies(size.x * size.y, Optional.empty()));
```
[the remark this answers is not in the document]
Comment (PS-Soundwave) [reply engine/src/main/java/org/terasology/world/generation/facets/base/BaseStrictlySparseFieldFacet2D.java]: Tell that to `SparseFieldFacet2D`, which is definitely not sparse at all.

This class was based off of `FieldFacet2D`, but could presumably be based on `SparseFieldFacet2D` with no default value as you suggest.

File: engine/src/main/java/org/terasology/logic/spawner/AbstractSpawner.java [hunk 3]
```diff
@@ -47,28 +52,59 @@ protected Vector3f findSpawnPosition(World world, Vector2i pos, int searchRadius
...
         for (BaseVector2i test : spiral) {
-            float val = surfaceHeightFacet.getWorld(test.getX(), test.getY());
-            int height = TeraMath.floorToInt(val);
-            if (height >= seaLevel) {
+            Optional<Float> val = getWorld.apply(test);
+            if(!val.isPresent()) {
+                continue;
+            }
+            int height = TeraMath.floorToInt(val.get());
+            if (!getSeaLevel.apply(test).isPresent() || height >= getSeaLevel.apply(test).get()) {
```
[the remark this answers is not in the document]
Comment (PS-Soundwave) [reply engine/src/main/java/org/terasology/logic/spawner/AbstractSpawner.java]: Do you mean
```java
if (getSeaLevel.apply(test).isPresent() && height >= getSeaLevel.apply(test).get() || !getSeaLevel.apply(test).isPresent()) {
```
?

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/logic/spawner/AbstractSpawner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jellysnake) [reply engine/src/main/java/org/terasology/logic/spawner/AbstractSpawner.java]: My bad, that would be correct.
Also comparing the two the existing version appears to be better than the updated version so this suggestion can be ignored

**Comment by jellysnake** [comment]
Changes LGTM and, pending a few sanity tests, should be mergable

**Pull request merged**